DOCS: Add handling of GIL to pure.rst

matusvalo opened the pull request.

matusvalo commented:
Two notes:

1. it seems that `@cython.gil` is not supported. I suppose issue should be created.
2. I avoiding mentioning parameter of cython.nogil (except workaround for `@cython.gil`) because it is compile time parameter which I think does not make sense in pure python mode since we does not have macros implemented in pure mode.

da-woods commented:
> I avoiding mentioning parameter of cython.nogil (except workaround for `@cython.gil`) because it is compile time parameter which I think does not make sense in pure python mode since we does not have macros implemented in pure mode.

I think the real value of the parameter is for fused types: to release the GIL for an array of C doubles, but keep it for an array of python objects for example. That should be possible in pure Python mode.

If the documentation doesn't make that use clear then it's possibly something we should improve

da-woods commented on docs/src/tutorial/pure.rst:
```diff
@@ -209,6 +209,27 @@ Here is an example of a :keyword:`cdef` function::
+    with cython.gil:
+        # code block with GIL acquired
+
+  .. Note:: Cython currently does not support ``@cython.gil`` decorator. ``@cython.nogil(False)`` should be used instead.
```
```suggestion
  .. Note:: Cython currently does not support the ``@cython.gil`` decorator. ``@cython.nogil(False)`` should be used instead.
```

matusvalo commented:
I have added conditional acquiring/releasing GIL to the documentation. Moreover, I fixed wrong comment about using `cython.nogil(False)` instead of `cython.gil`. I released that `cython.nogil(False`) != `cython.gil`.

scoder commented on docs/src/tutorial/pure.rst:
```diff
@@ -209,6 +209,44 @@ Here is an example of a :keyword:`cdef` function::
+    with cython.gil:
+        # code block with the GIL acquired
+
+  .. Note:: Cython currently does not support the ``@cython.gil`` decorator.
```
I think it should be called `@cython.with_gil`. Just naming it `@cython.gil` seems unclear. (We named the modifier "... with gil" for the same reason.)

matusvalo replied:
Me sense. I will update that part.

scoder commented on docs/src/tutorial/pure.rst:
```diff
@@ -209,6 +209,44 @@ Here is an example of a :keyword:`cdef` function::
+
+  .. Note:: Cython currently does not support the ``@cython.gil`` decorator.
+
+When either ``cython.nogil`` or ``cython.gil`` is used as a function, it accepts a parameter for conditional
+releasing or acquiring the GIL. The condition must be constant (at compile time)::
```
```suggestion
Both directives accept an optional boolean parameter for conditionally
releasing or acquiring the GIL. The condition must be constant (at compile time)::
```

scoder replied:
It would also be nice to have an actual usage example for something less trivial. Passing `True` and `False` makes it look useless.

matusvalo replied:
> EDIT: or mention at least fused types as a common use case, with a type based condition.

Added

matusvalo commented:
@scoder  I have implemented your comments.

scoder commented on docs/src/tutorial/pure.rst:
```diff
@@ -209,6 +209,45 @@ Here is an example of a :keyword:`cdef` function::
+  with cython.gil(True):
+      # code block with the GIL acquired
+
+The common use case of conditional acquiring and releasing of the GIL are fused types (see :ref:`gil_conditional`).
```
```suggestion
A common use case for conditionally acquiring and releasing the GIL are fused types
that allow different GIL handling depending on the specific type (see :ref:`gil_conditional`).
```

scoder commented:
Thanks
